auction.sympl

```diff
@@ -33,95 +40,273 @@ schema Auction:
+    existing_admin = cvm.storage.get(PUBLIC, AdminStatic, Identifier('admin'))
+    if isinstance(existing_admin, Admin):
+        admin_ka : str = existing_admin.id
+        tx_ka : str = cvm.tx.key_alias
+        return admin_ka == tx_ka
+    else:
+        return False
+
+
+###########################################
+# Premium Channel API and Implementations #
+###########################################
+
+# Creates the premium channel. No other channels may be created after this one.
+@clientside
+def create_premium_channel() -> None:
+    if check_admin():
+        existing_premium = cvm.storage.get(PUBLIC, PremiumChannelStatic, Identifier('premium'))
+        if isinstance(existing_premium, None):
+            premium = cvm.new_channel('PRM')
+            with PostTxArgs(PUBLIC):
+                _create_premium_channel(premium)
+        else:
+            cvm.error(f"Premium auction channel already created.")
+    else:
+        cvm.error(f"Only the current admin may perform this function.")
+
+@executable
+def _create_premium_channel(premium_channel: ChannelName) -> None:
+    if check_admin():
+        existing_premium = cvm.storage.get(PUBLIC, PremiumChannelStatic, Identifier('premium'))
+        if isinstance(existing_premium, None):
+            premium = PremiumChannel(channel_id=premium_channel, members=[cvm.tx.key_alias])
+            cvm.storage.put(Identifier('premium'), premium)
+            cvm.create_event(f'Premium Channel {premium_channel} Created', std.json(premium))
+        else:
+            cvm.error(f"Premium auction channel already created.")
+    else:
+        cvm.error(f"Only the current admin may perform this function.")
+
+
+#############################################
+# Add/Remove Member API and Implementations #
+#############################################
+
+# Adds a member to the premium channel, giving them access to create and bid on auctions there.
+@clientside
+def add_member(new_member: KeyAlias, channel: ChannelName) -> None:
+    if check_admin():
+        premium = cvm.storage.get(PUBLIC, PremiumChannelStatic, Identifier('premium'))
+        if isinstance(premium, PremiumChannel):
+            with PostTxArgs(PUBLIC):
+                _add_member(new_member)
+        else:
+            cvm.error(f"Channel {premium} is not the Premium channel.")
+    else:
+        cvm.error(f"Only the current admin may perform this function.")
+    
+@executable
+def _add_member(new_member : KeyAlias) -> None:
+    if check_admin():
+        premium = cvm.storage.get(PUBLIC, PremiumChannelStatic, Identifier('premium'))
+        if isinstance(premium, PremiumChannel):
+            nm_ka : str = new_member
+            for member in premium.members:
+                m_ka : str = member
+                if nm_ka == m_ka:
+                    cvm.error(f"{new_member} is already a member of the premium channel.")
+            cvm.send_key(premium.channel_id, new_member)
+            premium.members += [new_member]
```

Comment:
you never use this list, beside for updating it, why are you storing it?

Reply:
having an explicit membership is the biggest problem of the chat contract, do you really need it here?

Reply:
Ah good to know! I actually based this functionality off of the chat contract, not knowing that was a problem.

Reply:
@symbiont-riccardo-abbate Follow up questions on this: why is having the list of members a problem? In real life, would we want the admin to have to contact users externally to ask for their KAs and keep track of them manually? In the event of need to remove a member from the secure channel, what if that member refuses? What would the expected workflow be if we don't keep track of the list of KAs as part of the schema?

Reply:
well, the problem is our platform does not support having this list of members. You might know there is an issue going on with the chat contract right now. The issue is NOT caused by the contract, but the presence of this KA list made the issue pop up. The membership management is done at the txe level, not at the contract level, so what ever you do with your list does not matter to txe, and you cannot even make these two operations (contract level and txe level) atomic, so you are bound to have problems, unless you are careful enough to write the contract in a way to avoid ALL possible race conditions.
In real life, each application will have to solve this issue in the way that works with them.
There is no concept of refusing to to be removed.